### PR TITLE
[bugfix] Fix idempotency by removing dhcp_option_ids from [aci_dhcp_option_policy]

### PR DESCRIPTION
**Pull request by shrsr**

### CODE REVIEW

File: aci/resource_aci_dhcpoptionpol.go [hunk 1]
```diff
@@ -99,26 +94,25 @@ func getRemoteDHCPOptionPolicy(client *client.Client, dn string) (*models.DHCPOp
...
 	if dhcpOptionPol.DistinguishedName == "" {
 		return nil, fmt.Errorf("DHCPOptionPolicy %s not found", dhcpOptionPol.DistinguishedName)
 	}
-
 	return dhcpOptionPol, nil
 }
 
-func getRemoteDHCPOptionFromDHCPOptionPolicy(client *client.Client, dn string) (*models.DHCPOption, error) {
-	dhcpOptionCont, err := client.Get(dn)
+func getRemoteDHCPOptionFromDHCPOptionPolicy(client *client.Client, dn string) ([]*models.DHCPOption, error) {
```
Comment (lhercot) [inline aci/resource_aci_dhcpoptionpol.go:104]: getRemoteDHCPOptionFromDHCPOptionPolicy -> getRemoteDHCPOptionsFromDHCPOptionPolicy

File: aci/resource_aci_dhcpoptionpol.go [hunk 2]
```diff
@@ -99,26 +94,25 @@ func getRemoteDHCPOptionPolicy(client *client.Client, dn string) (*models.DHCPOp
...
 	}
 
-	dhcpOption := models.DHCPOptionFromContainer(dhcpOptionCont)
+	dhcpOption := models.DHCPOptionListFromContainer(dhcpOptionCont)
 
-	if dhcpOption.DistinguishedName == "" {
-		return nil, fmt.Errorf("DHCPOption %s not found", dhcpOption.DistinguishedName)
+	for _, option := range dhcpOption {
+		if option.DistinguishedName == "" {
+			return nil, fmt.Errorf("DHCPOption %s not found", option.DistinguishedName)
```
Comment (lhercot) [inline aci/resource_aci_dhcpoptionpol.go:114]: This error message tries to insert an empty string in the log

File: aci/resource_aci_dhcpoptionpol.go [hunk 3]
```diff
@@ -296,50 +292,47 @@ func resourceAciDHCPOptionPolicyUpdate(ctx context.Context, d *schema.ResourceDa
...
 	if d.HasChange("dhcp_option") {
-		dhcpOption := d.Get("dhcp_option_ids").([]interface{})
-		for _, val := range dhcpOption {
-			dhcpOptionDN := val.(string)
-			err := aciClient.DeleteByDn(dhcpOptionDN, "dhcpOption")
+		old_options, _ := d.GetChange("dhcp_option")
+
+		for _, optionParam := range old_options.([]interface{}) {
+			optionMap := optionParam.(map[string]interface{})
+			err := aciClient.DeleteByDn(optionMap["id"].(string), "dhcpOption")
```
Comment (lhercot) [inline aci/resource_aci_dhcpoptionpol.go:299]: Can we check if they are actually different before deleting them?

File: aci/resource_aci_dhcpoptionpol.go [hunk 4]
```diff
@@ -367,20 +360,13 @@ func resourceAciDHCPOptionPolicyRead(ctx context.Context, d *schema.ResourceData
...
-		dhcpOptions = append(dhcpOptions, dhcpOption)
-
-	}
-	_, err = setDHCPOptionAttributesFromDHCPOptionPolicy(dhcpOptions, d)
+	dhcpOptions, err := getRemoteDHCPOptionFromDHCPOptionPolicy(aciClient, dn)
 	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	_, err_option := setDHCPOptionAttributesFromDHCPOptionPolicy(dhcpOptions, d)
```
Comment (lhercot) [inline aci/resource_aci_dhcpoptionpol.go:368]: err_options or err_set_options

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/resource_aci_dhcpoptionpol.go (hunk 2 above)
Comment (shrsr) [reply aci/resource_aci_dhcpoptionpol.go:114]: Yeah, I'll change it.

File: aci/resource_aci_dhcpoptionpol.go (hunk 3 above)
Comment (shrsr) [reply aci/resource_aci_dhcpoptionpol.go:299]: if d.HasChange("dhcp_option") ensures that the option is different?

File: aci/resource_aci_dhcpoptionpol.go [hunk 5]
```diff
@@ -99,26 +95,25 @@ func getRemoteDHCPOptionPolicy(client *client.Client, dn string) (*models.DHCPOp
...
 	}
 
-	dhcpOption := models.DHCPOptionFromContainer(dhcpOptionCont)
+	dhcpOption := models.DHCPOptionListFromContainer(dhcpOptionCont)
 
-	if dhcpOption.DistinguishedName == "" {
-		return nil, fmt.Errorf("DHCPOption %s not found", dhcpOption.DistinguishedName)
+	for _, option := range dhcpOption {
+		if option.DistinguishedName == "" {
+			return nil, fmt.Errorf("DHCPOption %s not found", dn)
```
Comment (lhercot) [inline aci/resource_aci_dhcpoptionpol.go:115]: dn is the parent here right? this condition and error message does not really make sense.
Shouldn't it return that error when no DHCPOption are returned? so the lengh of the list is probably better.

File: aci/resource_aci_dhcpoptionpol.go [hunk 6]
```diff
@@ -99,26 +95,25 @@ func getRemoteDHCPOptionPolicy(client *client.Client, dn string) (*models.DHCPOp
...
-func getRemoteDHCPOptionFromDHCPOptionPolicy(client *client.Client, dn string) (*models.DHCPOption, error) {
-	dhcpOptionCont, err := client.Get(dn)
+func getRemoteDHCPOptionsFromDHCPOptionPolicy(client *client.Client, dn string) ([]*models.DHCPOption, error) {
+	dhcpOptionCont, err := client.GetViaURL(fmt.Sprintf("%s/%s/%s.json", "/api/node/class", dn, "dhcpOption"))
 	if err != nil {
 		return nil, err
 	}
 
-	dhcpOption := models.DHCPOptionFromContainer(dhcpOptionCont)
+	dhcpOption := models.DHCPOptionListFromContainer(dhcpOptionCont)
```
Comment (lhercot) [inline aci/resource_aci_dhcpoptionpol.go:111]: dhcpOptions

File: aci/resource_aci_dhcpoptionpol.go [hunk 7]
```diff
@@ -296,50 +293,55 @@ func resourceAciDHCPOptionPolicyUpdate(ctx context.Context, d *schema.ResourceDa
...
-		dhcpOption := d.Get("dhcp_option_ids").([]interface{})
-		for _, val := range dhcpOption {
-			dhcpOptionDN := val.(string)
-			err := aciClient.DeleteByDn(dhcpOptionDN, "dhcpOption")
-			if err != nil {
-				return diag.FromErr(err)
+		old_options, new_options := d.GetChange("dhcp_option")
+		getDifference := differenceInMaps(old_options.(*schema.Set), new_options.(*schema.Set))
+		listDNToDelete := make([]string, 0)
+		for i := 0; i < len(getDifference); i++ {
```
Comment (lhercot) [inline aci/resource_aci_dhcpoptionpol.go:299]: Why not use for ... := range ...  and use this for loop syntax?

File: aci/resource_aci_dhcpoptionpol.go [hunk 8]
```diff
@@ -296,50 +293,55 @@ func resourceAciDHCPOptionPolicyUpdate(ctx context.Context, d *schema.ResourceDa
...
-			dhcpOptionAttr := models.DHCPOptionAttributes{}
-			dhcpOption := val.(map[string]interface{})
-
-			name := dhcpOption["name"].(string)
+		for _, value := range listDNToDelete {
+			if value != "" {
+				err := aciClient.DeleteByDn(value, "dhcpOption")
+				if err != nil {
+					return diag.FromErr(err)
+				}
```
Comment (lhercot) [inline aci/resource_aci_dhcpoptionpol.go:312]: Any reason to do the delete in another loop and not directly in the previous one?

File: aci/resource_aci_dhcpoptionpol.go [hunk 9]
```diff
@@ -296,50 +293,55 @@ func resourceAciDHCPOptionPolicyUpdate(ctx context.Context, d *schema.ResourceDa
...
-			dhcpOptionDN := val.(string)
-			err := aciClient.DeleteByDn(dhcpOptionDN, "dhcpOption")
-			if err != nil {
-				return diag.FromErr(err)
+		old_options, new_options := d.GetChange("dhcp_option")
+		getDifference := differenceInMaps(old_options.(*schema.Set), new_options.(*schema.Set))
+		listDNToDelete := make([]string, 0)
+		for i := 0; i < len(getDifference); i++ {
+			for key, value := range getDifference[i].(map[string]interface{}) {
+				if key == "id" {
```
Comment (lhercot) [inline aci/resource_aci_dhcpoptionpol.go:301]: Do we have to loop through the keys to access a specific id key?

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/resource_aci_dhcpoptionpol.go (hunk 8 above)
Comment (shrsr) [reply aci/resource_aci_dhcpoptionpol.go]: This section is not required if I use for loop with range above.

File: aci/resource_aci_dhcpoptionpol.go (hunk 9 above)
Comment (shrsr) [reply aci/resource_aci_dhcpoptionpol.go]: The key of dn for dhcpoption is "id"

File: aci/resource_aci_dhcpoptionpol.go (hunk 7 above)
Comment (shrsr) [reply aci/resource_aci_dhcpoptionpol.go]: Changed the format!

File: aci/resource_aci_dhcpoptionpol.go (hunk 5 above)
Comment (shrsr) [reply aci/resource_aci_dhcpoptionpol.go]: Yes. I did change this part quite a bit. User should be able to add dhcp option policy with no options specified as well.

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**